ui/src/trace_processor/http_rpc_engine.ts

```diff
@@ -86,11 +88,20 @@ export class HttpRpcEngine extends EngineBase {
   }
 
   private onWebsocketMessage(e: MessageEvent) {
-    assertExists(e.data as Blob)
-      .arrayBuffer()
-      .then((buf) => {
-        super.onRpcResponseBytes(new Uint8Array(buf));
-      });
+    const blob = assertExists(e.data as Blob);
+    this.queue.push(blob);
+    this.processQueue();
+  }
+
+  private async processQueue() {
+    if (this.isProcessingQueue) return;
+    this.isProcessingQueue = true;
+    while (this.queue.length > 0) {
+      const blob = assertExists(this.queue.shift());
+      const buf = await blob.arrayBuffer();
+      super.onRpcResponseBytes(new Uint8Array(buf));
```

Comment:
If anything throws inside this handler could we end up leaving ourselves in an invalid state?